### DISCUSSION TITLE
p146 UpdateTodo 메서드 질문

### DISCUSSION THREAD

**Discussion by devbelly** [fsoftwareengineer/todo-application#44] [질문과 답변] [answered]
작가님 안녕하세요. 공부를 하다가 모르는 부분이 있어 질문을 남깁니다.

1. p146 UpdateTodo 메서드는 왜 createTodo, deleteTodo와는 다르게 try-catch 블록을 사용하지 않은 것인가요? service.update(entity)에서 entity에 대한 validation을 진행하므로 exception이 있을 수 있는데 try-catch를 사용하지 않은 이유가 궁금합니다.

2. p 149의 실습코드 2-40을 포함하여 TodoService클래스의 메서드 작성 시, 메서드 인자로 final을 사용하는데, final을 사용해야하는 이유가 궁금합니다.

좋은 책을 써 주셔서 항상 감사합니다.

**Comment by fsoftwareengineer** [comment, marked as answer]
1. UpdateTodo에도 try-catch가 있는것이 좋습니다. 책에서는 에러처리에 큰 비중을 두지 않아 누락된 것 같습니다.
2. 메서드의 인자로 `final`을 사용하는 것은 coding convention(자바 프로그래머들이 코드를 작성할때 따르는 보편적인 규칙)의 일종입니다. 다시말하면 final을 굳이 사용하지 않아도 됩니다. 그러나 보통 final을 사용하는 이유는 개발자들이 메서드 인자를 수정하는 것을 지양하기 때문입니다. 메서드 인자를 수정하면 버그가 나기 쉬워서 그런데요, 여기서는 하나의 예만 들도록 하겠습니다.

예를 들어 아래와같은 코드를 생각 해 볼 수 있습니다.
```

public class Main {

    public static void main(String[] args) {
        Main m = new Main();
        MyObject o1 = null;
        m.getObject(o1);
        System.out.println(o1.toString());
    }

    public void getObject(MyObject o) {
        o = new MyObject("World");
    }

}

```

이 때 System.out.println의 결과는 무엇일까요? 바로 NullPointerException입니다. getObject안에서 초기화 해 주었는데 왜 안될까요? 메서드가 한번 호출되고 나면 내부에서 사용된 모든 로컬 변수는 사라집니다.  MyObject o는 밖에서 넘어온 로컬변수(=매개변수) 입니다. 따라서 로컬 변수인 o에 새 오브젝트를 할당해도 메서드가 끝나면 참조하는 다른 변수가 없으므로 사라집니다. 매개변수인 o는 o1에서 넘어 온 것인데 왜 그렇게 될까요? 이는 자바가 call by value라는 매개변수 평가전략을 사용하기 때문입니다. 더 자세한 내용은 https://imasoftwareengineer.tistory.com/104 에서 확인 할 수 있습니다.

이런식으로 매개변수 자체를 계속 변경하다보면 큰 프로젝트에서는 하나의 오브젝트가 어디서 생성된고 어디서 변경되는지 알기 힘들어집니다. 컴파일 당시에 에러가 나지 않기 때문에 버그를 잡는 것도 쉽지 않습니다. 따라서 보통 개발자들은 매개변수 또는 일부 로컬변수에 final을 붙여 해당 변수를 수정하지 못하도록 합니다.
